src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -336,20 +279,128 @@ ErrorIfUnsupportedForeignConstraintExists(Relation relation, char referencingDis
 		 * placements always be in the same state (b) executors are aware of reference
 		 * tables and handle concurrency related issues accordingly.
 		 */
-		if (!referencingNotReplicated)
+		EnsureReferencingTableNotReplicated(referencingTableId);
+
+		ReleaseSysCache(heapTuple);
+	}
+}
+
+
+/*
+ * EnsureSupportedFKeyBetweenCitusLocalAndRefTable is an helper function that
+ * takes a foreign key contraint form for a foreign key between two
+ * DISTRIBUTE_BY_NONE tables and errors out if it it an unsupported foreign key
+ * from a reference table to a citus local table according to given replication
+ * model parameters.
+ */
+static void
+EnsureSupportedFKeyBetweenCitusLocalAndRefTable(Form_pg_constraint fKeyConstraintForm,
+												char referencingReplicationModel,
+												char referencedReplicationModel)
+{
+	bool referencingIsReferenceTable =
+		(referencingReplicationModel == REPLICATION_MODEL_2PC);
+	bool referencedIsCitusLocalTable =
+		(referencedReplicationModel != REPLICATION_MODEL_2PC);
+	if (referencingIsReferenceTable && referencedIsCitusLocalTable)
+	{
+		/*
+		 * We only support RESTRICT and NO ACTION behaviors for the
+		 * foreign keys from reference tables to citus local tables.
+		 * This is because, we can't cascade dml operations from citus
+		 * local tables's coordinator placement to the remote placements
+		 * of the reference table.
+		 * Note that for the foreign keys from reference tables to citus
```

Comment:
Should this be `from citus local tables to reference tables` ?

Reply:
good catch, thanks